### PR TITLE
Initial OpenStack external CCM setup.

### PR DESCRIPTION
**Pull request by ainmosni**
**What this PR does / why we need it**:
Adds support for the external OpenStack cloud controll manager.

As the internal OpenStack code in the controller manager
will be gone soon, we'll need to switch to the external cloud controller
manager, this PR adds the configuration for it.

```release-note
* Adds support for the external OpenStack Cloud Control Manager.
```

### CODE REVIEW

**Comment by xmudrii** [comment]
Hello @ainmosni,

thank you for the PR! However, before we can merge this, there are several things we need to take care of and decide about. The switch to the external CCMs is a little bit tough topic because there are some points that are still unclear.

You can find some of the discussion in #616

**Comment by ainmosni** [comment]
@xmudrii Yeah the migration will be painful but as there's no real development on the in-tree one and the in-tree code will be removed sooner rather later, it might be worth to pull the bandaid off sooner rather than later. I can definitely take a look at the OpenStack side to help with the migration options.

**Comment by kron4eg** [comment]
> @xmudrii Yeah the migration will be painful but as there's no real development on the in-tree one and the in-tree code will be removed sooner rather later

AFAIK, timeline is set to 1.17.x

>, it might be worth to pull the bandaid off sooner rather than later.

Indeed, that's why we create #616.

>I can definitely take a look at the OpenStack side to help with the migration options.

There is a problem with openstack CCM. Its releases are bound to specific minor kubernetes versions. So it's not that easy to jump on it and we need to take into account older version (since we support k8s starting from 1.13).

**Comment by ainmosni** [comment]
Yeah I kept that in mind in this patch, AFAICS they only started to tag their images since 1.15, they used `latest` before that.

One thing that I didn't put in this patch but we should probably add is support for the cinder CSI plugin, so that PVCs keep working.

**Comment by kron4eg** [comment]
OK, I've got a confirmation that it's safe to use OS CCM 1.15 in 1.13, 1.14 and 1.15 clusters (but no 1.12, but since we don't support 1.12 anyway, we are safe).

File: pkg/templates/externalccm/openstack.go [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package externalccm
+
+import (
```
Comment (kron4eg) [inline pkg/templates/externalccm/openstack.go]: please take a look at https://github.com/kubermatic/kubeone/blob/master/CONTRIBUTING.md#import-order

**Comment by ainmosni** [comment]
@kron4eg Yeah that's what I gathered from their examples as well. It's a bit more complicated with cinder-csi (which isn't in this PR) as the CSI sidecar containers are k8s version specific. I was thinking about adding those, but changed my mind to keep the PR smaller.

File: pkg/templates/externalccm/openstack.go (hunk 1 above)
Comment (ainmosni) [reply pkg/templates/externalccm/openstack.go]: Sorted. :)

**Comment by ivomarino** [comment]
Hi all, is this PR going to happen "soon" or what's the best actual solution on this topic? Thanks

**Comment by kron4eg** [comment]
@ivomarino yes, in the nearest future, I'd say. I'm working on integrating changes from this PR.

**Comment by kron4eg** [comment]
This PR is replaced by https://github.com/kubermatic/kubeone/pull/820

@ainmosni thanks for initiating this!

**Pull request closed without merging**